myhoard/controller.py

```diff
@@ -1231,7 +1255,14 @@ def _purge_old_backups(self):
         # is called repeatedly so if there are for any reason more backups
         # to drop they will be dropped soon enough
         purgeable = sort_completed_backups(purgeable)
+        broken_backups = sum(backup["broken_at"] is not None for backup in purgeable)
         backup = purgeable[0]
+
+        if backup["broken_at"] is None and broken_backups and broken_backups == len(purgeable) - 1:
```

Comment:
I admit I'm not sure I understand this logic, and since I don't understand the test name, I'm left a bit in the dark. What is the outcome you desire? The idea is that if some backup is corrupted, but this one isn't, we don't delete it, right?

Backups are sorted left-to-right in decreasing age order (i.e. backups[-1] is the most recent, backups[0] is the oldest). When purging, we usually try to purge the oldest backup first (hence the purgeable[0]) above here.

IMHO (but verify, I may be missing some piece) the ideal logic would be:
- if there's any broken backup in the chain, purge the oldest broken backup and leave all non-broken backups there;
- if there're no broken backups, just purge the oldest.

Reply:
> if there's any broken backup in the chain, purge the oldest broken backup

We store binlogs per basebackup and while the basebackup may be broken the binlogs associated with it are probably valid. Deleting a backup from middle, broken or no, would leave a hole in binlogs and break our ability to restore to any point in time since the oldest backup and we should not do that.

Reply:
Thanks Rauli! OK, this makes sense to me (I was missing some piece after all) . So we should either 

a) not alter the purging logic at all; this is quite safe, and the restore will pick the correct basebackup.

or

b) try purging old, broken basebackups, if the oldest is broken, e.g.

if we have basebackups T1, T2, T3, T4, T5; with T1 being the oldest; if T1, T2 and T4 are broken, we can safely delete T1 and T2 (and all dependent binglogs) as well as delete them, since there's no way to restore them or anything that depends on them, while, OTC, we should retain T4 since we can still restore from T3 and use the T4 binlogs.

BUT:
I'd probably not enable this logic in production from the beginning; if we accidentally mark a basebackup as broken, we may end up deleting too many basebackups. Probably I'd add the logic with just a metric/logline, to see if it can really happen.

Reply:
as discussed during daily, in case the oldest backup is the only non-broken, it will be only purged when a new backup is generated